encoding/src/serde_byte_array.rs

```diff
@@ -5,48 +5,41 @@ use crate::BYTE_ARRAY_MAX_LEN;
 use serde::{de, ser, Deserializer, Serializer};
 use serde_bytes::{Deserialize, Serialize};
 
-/// serde_bytes with max length check
-pub mod serde_byte_array {
-    use super::*;
+/// checked if input > `crate::BYTE_ARRAY_MAX_LEN`
+pub fn serialize<T, S>(bytes: &T, serializer: S) -> Result<S::Ok, S::Error>
+where
+    T: ?Sized + Serialize + AsRef<[u8]>,
+    S: Serializer,
+{
+    if bytes.as_ref().len() > BYTE_ARRAY_MAX_LEN {
+        return Err(ser::Error::custom::<String>(
+            "Array exceed max length".into(),
+        ));
+    }
+
+    Serialize::serialize(bytes, serializer)
+}
 
-    /// checked if input > `crate::BYTE_ARRAY_MAX_LEN`
-    pub fn serialize<T, S>(bytes: &T, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        T: ?Sized + Serialize + AsRef<[u8]>,
-        S: Serializer,
-    {
-        let len = bytes.as_ref().len();
-        if len > BYTE_ARRAY_MAX_LEN {
-            return Err(ser::Error::custom::<String>(
+/// checked if output > `crate::ByteArrayMaxLen`
+pub fn deserialize<'de, T, D>(deserializer: D) -> Result<T, D::Error>
+where
+    T: Deserialize<'de> + AsRef<[u8]>,
+    D: Deserializer<'de>,
+{
+    Deserialize::deserialize(deserializer).and_then(|bytes: T| {
+        if bytes.as_ref().len() > BYTE_ARRAY_MAX_LEN {
```

Comment:
Perhaps it would be useful to add a debug log indicating what was the size of the offending array. Or perhaps put it into the error message? What do you think?

Reply:
agree with this idea

tried to place this error into https://github.com/ChainSafe/forest/blob/main/encoding/src/errors.rs but it seems that's not where it should be lol

adding a new `Error` type is kind of luxury for this one line error, so I just placed a string which can be captured by `serde` for the minimal modification against to the last commit

or I can add an `Any` item in [`CodecProtocol`](https://github.com/ChainSafe/forest/blob/e66d0f9e2f810752bd3330d9ae2370f7197b234b/encoding/src/errors.rs#L50) to describe this error

```rust
pub enum CodecProtocol {
  // use any since this error occurs in serialize && deserialize
  //
  // add this line ⬇️
  Any,
  Cbor,
}
```

and the error would be like

```rust
let arr_limit_err = encoding::Error {
  description: "Array exceeds max length".into(),
  protocal: CodecProtocol::Any,
}
```

Reply:
Agree with @clearloop. It's kind of luxury, in other parts of the codebase we just `format!` the string with the interesting value, e.g. [here](https://github.com/ChainSafe/forest/blob/main/utils/bitfield/src/rleplus/mod.rs#L100).

Reply:
Sure, if you think it's an overkill then it's fine to me. It potentially makes error matching more difficult later on but I'm new to the codebase, if it's unlikely and the error rarely occurs then indeed it may be too much.